src/Data/List/Lazy.purs

```diff
@@ -757,6 +759,17 @@ foldrLazy op z = go
       Cons x xs' -> Z.defer \_ -> x `op` go xs'
       Nil -> z
 
+-- | Perform a right fold until the list is consumed or the provided function returns a Left value
+foldrWhile :: forall a b. (a -> b -> Either b b) -> b -> List a -> b
+foldrWhile f acc xs = go $ step xs
+  where
+    go :: Step a -> b
+    go Nil = acc
+    go (Cons x xs') =
+      case f x acc of
```

Comment:
I could use [either](https://pursuit.purescript.org/packages/purescript-either/4.1.1/docs/Data.Either#v:either) here, but this feels more obvious to me.